basis/basis.py

```diff
@@ -0,0 +1,172 @@
+    constructing a surrogate function of any accuracy at or below 
+    max_exactness.
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid. The Smolyak index of a given point is equal to 
+    its position in the list points.
+    ``levels`` property gives the Smolyak indexes for each
+    grid level.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
+        self._max_exactness = max_exactness
+        self._points = []
+        self._levels = []
+        self._update()
+
+    @property
+    def max_exactness(self):
+        """Maximum level of exactness described"""
+        return self._max_exactness
+
+    @property
+    def points(self):
+        """Extrema assigned to Smolyak indices"""
+        return self._points
+
+    @property
+    def levels(self):
+        """index of each point that belongs to each grid level"""
+        return self._levels
+
+    @max_exactness.setter
+    def max_exactness(self, max_exactness):
+        """Set maximum level of exactness
+        Sets maximmum level of exactness class instance will describe, will
+        extend or truncate properties points and levels to match new value
+
+        Parameters
+        ----------
+        max_exactness : int
+            new level of exactness class should describe
+        """
+        if max_exactness > self._max_exactness:
+            # extend
+            self._max_exactness = max_exactness
+            # update other properties
+            self._update()
+        elif max_exactness < self._max_exactness:
+            # truncate
+            self._max_exactness = max_exactness
+            # update other properties
+            points_keep = sum(
+                    [len(x) for x in self._levels[:self._max_exactness+1]])
+            self._points = self._points[:points_keep]
+            self._levels = self._levels[:self._max_exactness+1]
+
+
+    @abc.abstractmethod
+    def _update(self):
+        """Update properties described by level of exactness
+        Compute points of basis function and the Smolyak index of
+        each grid level between 0 and max_exactness + 1 
+        """
+        pass
+
+    @abc.abstractmethod
+    def basis(self,x,n):
+        """Compute term of basis function
+        Returns the output of an nth degree basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        n : int
+            degree
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis Function of Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a set of polynomials described by the
+    following recurrence relation:
```

Comment:
This documentation is a bit unclear to me: I think that this is actually a family of basis functions, right? (See comment above on whether `__call__` should include `n`)